lib/schedule/minischedule/load.ex

```diff
@@ -35,45 +37,97 @@ defmodule Schedule.Minischedule.Load do
 
   @spec run(Run.key(), [Activity.t()] | nil, [Trip.t()] | nil) :: Run.t()
   def run(run_key, activities, trips) do
+    {schedule_id, run_id} = run_key
     activities = activities || []
     trips = trips || []
-    # TODO real implementation.
-    # Currently, turns the trips directly into pieces, and makes every activity a break.
-    # The real way to do it would be to integrate Sign-on and Operator activities into pieces.
-    {schedule_id, run_id} = run_key
 
-    breaks = Enum.map(activities, &Break.from_activity/1)
-
-    pieces =
-      trips
-      |> Enum.group_by(fn trip -> trip.block_id end)
-      |> Enum.map(fn {block_id, trips} ->
-        first_trip = List.first(trips)
-        last_trip = List.last(trips)
-
-        %Piece{
-          schedule_id: schedule_id,
-          run_id: run_id,
-          block_id: block_id,
-          start: %{
-            time: first_trip.start_time,
-            place: first_trip.start_place,
-            mid_route?: false
-          },
-          trips: Enum.map(trips, fn trip -> trip.trip_id end),
-          end: %{
-            time: last_trip.end_time,
-            place: last_trip.end_place,
-            mid_route?: false
-          }
-        }
-      end)
+    activities =
+      activities
+      |> operators_to_pieces(trips)
+      |> add_sign_ons_to_pieces()
+      |> activities_to_breaks()
 
     %Run{
       schedule_id: schedule_id,
       id: run_id,
-      activities: breaks ++ pieces
+      activities: activities
+    }
+  end
+
+  @spec operators_to_pieces([Activity.t()], [Trip.t()]) :: [Activity.t() | Piece.t()]
+  defp operators_to_pieces([], trips) do
+    if !Enum.empty?(trips) do
+      Logger.warn("Some trips not matched", trips: trips)
+    end
+
+    []
+  end
+
+  defp operators_to_pieces([activity | rest], trips) do
+    case activity.activity_type do
+      "Operator" ->
+        {piece, remaining_trips} = operator_to_piece(activity, trips)
+        [piece | operators_to_pieces(rest, remaining_trips)]
+
+      _ ->
+        # non-Operator activities are untouched
+        [activity | operators_to_pieces(rest, trips)]
+    end
+  end
+
+  @spec operator_to_piece(Activity.t(), [Trip.t()]) :: {Piece.t(), [Trip.t()]}
+  defp operator_to_piece(%Activity{activity_type: "Operator"} = activity, trips) do
```

Comment:
This is less of a case and more of an assertion that this function should only be called on Operator activities. (Same thing below in `trip_in_operator`.)

The recursion in `operators_to_pieces` is because that function works its way through both the list of trips and the list of activities, which can't be done by a map.

If the matching trips were found with a `Enum.filter(trips)` instead of `Enum.split_while(trips)`, it would mean that this could be a simple map, but we'd lose the ability to detect if there are any unmatched trips.

I guess now that this has gone checked without detecting any unmatched trips, it seems like the data should always be consistent, and it'd be okay to do the simpler solution. I'll make that change.

Reply:
Whatever seems best to you. I think I realized the need for recursion for the trips later on in my review and forgot to delete the comment.